# Extracción, Transformación y Carga (ETL)

En esta sección, daré inicio al proceso de Extracción de Datos con el objetivo de verificar la naturaleza de los datos con los que trabajaré. Posteriormente, llevaré a cabo un análisis detallado y realizaré las métricas asignadas. (El Analisis va enfocado a los Automoviles de combustible)

In [77]:
# Importo las librerias necesarias.
import pandas as pd
import numpy as np
import datetime

In [2]:
# Lectura del primer dataframe 'yellow_tripdata_2023-09.parquet'.
# Y visualizacion de las primeras 15 filas.
df= pd.read_parquet('../parquet/yellow_tripdata_2023-09.parquet')
df.head(15)


VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0          1  2023-09-01 00:15:37   2023-09-01 00:20:21              1.0   
1          2  2023-09-01 00:18:40   2023-09-01 00:30:28              2.0   
2          2  2023-09-01 00:35:01   2023-09-01 00:39:04              1.0   
3          2  2023-09-01 00:45:45   2023-09-01 00:47:37              1.0   
4          2  2023-09-01 00:01:23   2023-09-01 00:38:05              1.0   
5          2  2023-09-01 00:05:20   2023-09-01 00:46:57              3.0   
6          1  2023-09-01 00:51:50   2023-09-01 01:10:21              0.0   
7          1  2023-09-01 00:01:04   2023-09-01 00:18:36              1.0   
8          2  2023-09-01 00:37:44   2023-09-01 00:52:24              2.0   
9          2  2023-09-01 00:02:13   2023-09-01 00:07:12              1.0   
10         2  2023-09-01 00:30:18   2023-09-01 00:58:36              1.0   
11         2  2023-09-01 00:08:32   2023-09-01 00:15:58              2.0   
12         2  2023-09-01 00:03:56   2023-09-01 00:23:09              1.0   
13         2  2023-09-01 00:40:28   2023-09-01 00:53:42              1.0   
14         2  2023-09-01 00:57:50   2023-09-01 01:02:37              2.0   

    trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0            0.80         1.0                  N           163           230   
1            2.34         1.0                  N           236           233   
2            1.62         1.0                  N           162           236   
3            0.74         1.0                  N           141           229   
4            9.85         1.0                  N           138           230   
5           12.83         1.0                  N            93            75   
6           10.90         1.0                  N            93           255   
7            3.90         1.0                  N           140             7   
8            2.70         1.0                  N            45           164   
9            1.02         1.0                  N           238           236   
10          16.66         2.0                  N           132           140   
11           1.37         1.0                  N           138           138   
12           8.17         1.0                  N           138           170   
13           2.31         1.0                  N           230           140   
14           0.95         1.0                  N           234           186   

    payment_type  fare_amount  extra  mta_tax  tip_amount  tolls_amount  \
0              2          6.5    3.5      0.5        0.00          0.00   
1              1         14.2    1.0      0.5        2.00          0.00   
2              1          8.6    1.0      0.5        2.00          0.00   
3              1          5.1    1.0      0.5        1.00          0.00   
4              1         45.0    6.0      0.5       17.02          0.00   
5              2         55.5    1.0      0.5        0.00          6.94   
6              1         41.5    1.0      0.5        3.00          0.00   
7              1         20.5    3.5      0.5        6.35          0.00   
8              1         16.3    1.0      0.5        4.26          0.00   
9              1          7.9    1.0      0.5        2.58          0.00   
10             2         70.0    0.0      0.5        0.00          0.00   
11             2          9.3    6.0      0.5        0.00          0.00   
12             1         33.1    6.0      0.5        8.00          6.94   
13             1         14.9    1.0      0.5        3.98          0.00   
14             1          7.2    1.0      0.5        2.44          0.00   

    improvement_surcharge  total_amount  congestion_surcharge  Airport_fee  
0                     1.0         11.50                   2.5         0.00  
1                     1.0         21.20                   2.5         0.00  
2                     1.0         15.60                   2.5         0.00

In [3]:
# Hago esta funcion para contar los datos faltantes en cada columna.
def contar_datos_faltantes(df):
    # Contar el número de datos faltantes en cada columna
    datos_faltantes = df.isnull().sum()
    return datos_faltantes

contar_datos_faltantes(df)

VendorID                      0
tpep_pickup_datetime          0
tpep_dropoff_datetime         0
passenger_count          140225
trip_distance                 0
RatecodeID               140225
store_and_fwd_flag       140225
PULocationID                  0
DOLocationID                  0
payment_type                  0
fare_amount                   0
extra                         0
mta_tax                       0
tip_amount                    0
tolls_amount                  0
improvement_surcharge         0
total_amount                  0
congestion_surcharge     140225
Airport_fee              140225
dtype: int64

In [4]:
df.info() # Muestra el tipo de dato de cada columna.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2846722 entries, 0 to 2846721
Data columns (total 19 columns):
 #   Column                 Dtype         
---  ------                 -----         
 0   VendorID               int32         
 1   tpep_pickup_datetime   datetime64[us]
 2   tpep_dropoff_datetime  datetime64[us]
 3   passenger_count        float64       
 4   trip_distance          float64       
 5   RatecodeID             float64       
 6   store_and_fwd_flag     object        
 7   PULocationID           int32         
 8   DOLocationID           int32         
 9   payment_type           int64         
 10  fare_amount            float64       
 11  extra                  float64       
 12  mta_tax                float64       
 13  tip_amount             float64       
 14  tolls_amount           float64       
 15  improvement_surcharge  float64       
 16  total_amount           float64       
 17  congestion_surcharge   float64       
 18  Airport_fee           

* En esta funcion ademas de calcular la duracion de un viaje por velocidad creo una nueva Columna llamada 'Duracion' que es por minutos.

In [5]:
# Hago esta funcion para calcular la duracion del viaje.
def calcular_duracion(df, tpep_pickup_datetime, tpep_dropoff_datetime):
    
    # Uso la funcion to_datetime para convertir las columnas en formato fecha.
    df[tpep_pickup_datetime] = pd.to_datetime(df[tpep_pickup_datetime])
    df[tpep_dropoff_datetime] = pd.to_datetime(df[tpep_dropoff_datetime])
    
    # Calcular la duración
    df['Duracion'] = df[tpep_dropoff_datetime] - df[tpep_pickup_datetime]
    
    return df

calcular_duracion(df, 'tpep_pickup_datetime', 'tpep_dropoff_datetime')

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0               1  2023-09-01 00:15:37   2023-09-01 00:20:21              1.0   
1               2  2023-09-01 00:18:40   2023-09-01 00:30:28              2.0   
2               2  2023-09-01 00:35:01   2023-09-01 00:39:04              1.0   
3               2  2023-09-01 00:45:45   2023-09-01 00:47:37              1.0   
4               2  2023-09-01 00:01:23   2023-09-01 00:38:05              1.0   
...           ...                  ...                   ...              ...   
2846717         2  2023-09-30 23:31:12   2023-09-30 23:48:29              NaN   
2846718         1  2023-09-30 23:42:18   2023-09-30 23:47:45              NaN   
2846719         1  2023-09-30 23:03:35   2023-09-30 23:14:50              NaN   
2846720         2  2023-09-30 23:57:05   2023-10-01 00:17:36              NaN   
2846721         1  2023-09-30 23:25:47   2023-09-30 23:55:28              NaN   

         trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
0                 0.80         1.0                  N           163   
1                 2.34         1.0                  N           236   
2                 1.62         1.0                  N           162   
3                 0.74         1.0                  N           141   
4                 9.85         1.0                  N           138   
...                ...         ...                ...           ...   
2846717           2.43         NaN               None           125   
2846718           0.00         NaN               None           236   
2846719           1.80         NaN               None           211   
2846720           3.39         NaN               None           209   
2846721           0.00         NaN               None            80   

         DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
0                 230             2         6.50    3.5      0.5        0.00   
1                 233             1        14.20    1.0      0.5        2.00   
2                 236             1         8.60    1.0      0.5        2.00   
3                 229             1         5.10    1.0      0.5        1.00   
4                 230             1        45.00    6.0      0.5       17.02   
...               ...           ...          ...    ...      ...         ...   
2846717           107             0        17.69    0.0      0.5        4.34   
2846718            75             0        11.33    0.0      0.5        0.00   
2846719            90             0        12.10    1.0      0.5        2.57   
2846720            97             0        20.33    0.0      0.5        4.87   
2846721           197             0        36.47    0.0      0.5        0.00   

         tolls_amount  improvement_surcharge  total_amount  \
0                 0.0                    1.0         11.50   
1                 0.0                    1.0         21.20   
2                 0.0                    1.0         15.60   
3                 0.0                    1.0         11.10   
4                 0.0                    1.0         73.77   
...               ...                    ...           ...   
2846717           0.0                    1.0         26.03   
2846718           0.0                    1.0         15.33   
2846719           0.0                    1.0         19.67   
2846720           0.0                    1.0         29.20   
2846721           0.0                    1.0         37.97   

         congestion_surcharge  Airport_fee        Duracion  
0                         2.5         0.00 0 days 00:04:44  
1                         2.5         0.00 0 days 00:11:48  
2                         2.5         0.00 0 days 00:04:03  
3                         2.5         0.00 0 days 00:01:52  
4                         2.5         1.75 0 days 00:36:42  
...                       ...          ...             ...  
2846717                   NaN          NaN 0 days 

El código proporcionado registra la duración en minutos del trayecto en un taxi de color amarillo, desde el punto de inicio hasta la conclusión del recorrido.

In [6]:
# Seleccionar solo las columnas 'tpep_pickup_datetime', 'tpep_dropoff_datetime' y 'Duracion'
df_seleccionado = df[['tpep_pickup_datetime', 'tpep_dropoff_datetime', 'Duracion']]

df_seleccionado.head()

tpep_pickup_datetime tpep_dropoff_datetime        Duracion
0  2023-09-01 00:15:37   2023-09-01 00:20:21 0 days 00:04:44
1  2023-09-01 00:18:40   2023-09-01 00:30:28 0 days 00:11:48
2  2023-09-01 00:35:01   2023-09-01 00:39:04 0 days 00:04:03
3  2023-09-01 00:45:45   2023-09-01 00:47:37 0 days 00:01:52
4  2023-09-01 00:01:23   2023-09-01 00:38:05 0 days 00:36:42

* Sección de Funciones para Calcular las Distancias de Vehículos con Combustible por Pasajero.<br>

Los taxis amarillos en Nueva York están autorizados para recoger más de un pasajero en un solo viaje, la tabla impresa nos muestra cuanto tardo el viaje y cuanto genero de ingresos.

In [7]:
def calcular_tiempo_por_distancia(df):
    
    df['Duracion'] = df['Duracion'].dt.total_seconds() / 60

    # Calcular la distancia recorrida por minuto
    df['Distancia_por_Minuto'] = df['trip_distance'] / df['Duracion']
    
    return df

df = calcular_tiempo_por_distancia(df)
df.head()

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0         1  2023-09-01 00:15:37   2023-09-01 00:20:21              1.0   
1         2  2023-09-01 00:18:40   2023-09-01 00:30:28              2.0   
2         2  2023-09-01 00:35:01   2023-09-01 00:39:04              1.0   
3         2  2023-09-01 00:45:45   2023-09-01 00:47:37              1.0   
4         2  2023-09-01 00:01:23   2023-09-01 00:38:05              1.0   

   trip_distance  RatecodeID store_and_fwd_flag  PULocationID  DOLocationID  \
0           0.80         1.0                  N           163           230   
1           2.34         1.0                  N           236           233   
2           1.62         1.0                  N           162           236   
3           0.74         1.0                  N           141           229   
4           9.85         1.0                  N           138           230   

   payment_type  ...  extra  mta_tax  tip_amount  tolls_amount  \
0             2  ...    3.5      0.5        0.00           0.0   
1             1  ...    1.0      0.5        2.00           0.0   
2             1  ...    1.0      0.5        2.00           0.0   
3             1  ...    1.0      0.5        1.00           0.0   
4             1  ...    6.0      0.5       17.02           0.0   

   improvement_surcharge  total_amount  congestion_surcharge  Airport_fee  \
0                    1.0         11.50                   2.5         0.00   
1                    1.0         21.20                   2.5         0.00   
2                    1.0         15.60                   2.5         0.00   
3                    1.0         11.10                   2.5         0.00   
4                    1.0         73.77                   2.5         1.75   

    Duracion  Distancia_por_Minuto  
0   4.733333              0.169014  
1  11.800000              0.198305  
2   4.050000              0.400000  
3   1.866667              0.396429  
4  36.700000              0.268392  

[5 rows x 21 columns]

En el análisis de este código, se destaca que la distancia mínima registrada es de 2.17 millas, con una duración de 14.28 minutos. Al calcular la relación entre la distancia y la duración, se obtiene un promedio de 0.001520 minutos por cada milla.

Por otro lado, se observa que la distancia máxima alcanzada es de  29.70 millas, con una duración de 62.86 minutos. En este caso, la tasa de distancia por cada minuto se estima en 0.47 millas.

In [34]:
def filtrar_viajes_largos_y_calcular_relacion(df):
    # Filtrar los viajes que duraron más de una hora
    df_viajes_largos = df[df['Duracion'] > 60]

    # Calcular la relación de distancia y tiempo
    df_viajes_largos.loc[:,'Distancia_por_Minuto'] = df_viajes_largos['trip_distance'] / df_viajes_largos['Duracion']
    
    return df_viajes_largos

df_viajes_largos = filtrar_viajes_largos_y_calcular_relacion(df[['trip_distance', 'Duracion', 'Distancia_por_Minuto']])
df_viajes_largos.head()

trip_distance     Duracion  Distancia_por_Minuto
714           25.71    67.433333              0.381265
1063          14.02    71.150000              0.197048
2239          29.70    62.866667              0.472428
2841           2.17  1427.766667              0.001520
3116          17.63    94.033333              0.187487

En esta otra funcion de viajes medios de taxis amarillos se destaca que la distancia mínima registrada es de 6.70 millas, con una duración de 30.40 minutos. Al calcular la relación entre la distancia y la duración, se obtiene un promedio de 0.22 minutos por cada milla.

Por otro lado, se observa que la distancia máxima alcanzada es de 24.81 millas, con una duración de 44.26 minutos. En este caso, la tasa de distancia por cada minuto se estima en 0.56 millas.

In [35]:
def filtrar_viajes_medios_y_calcular_relacion(df):
    # Filtrar los viajes que duraron más de una hora
    df_viajes_medios = df[df['Duracion'] >= 30]

    # Calcular la relación de distancia y tiempo
    df_viajes_medios.loc[:,'Distancia_por_Minuto'] = df_viajes_medios['trip_distance'] / df_viajes_medios['Duracion']
    
    return df_viajes_medios

df_viajes_medios = filtrar_viajes_medios_y_calcular_relacion(df[['trip_distance', 'Duracion', 'Distancia_por_Minuto']])
df_viajes_medios.head()

trip_distance   Duracion  Distancia_por_Minuto
4            9.85  36.700000              0.268392
5           12.83  41.616667              0.308290
15          12.20  44.866667              0.271917
42           6.70  30.400000              0.220395
43          24.81  44.266667              0.560467

En esta otra funcion de viajes cortos de taxis amarillos, la distancia mínima registrada es de 0.74 millas, con una duración de 1.86 minutos. Al calcular la relación entre la distancia y la duración, se obtiene un promedio de 0.39 minutos por cada milla.

Por otro lado, se observa que la distancia máxima alcanzada es de 1.62 millas, con una duración de 4.05 minutos. y la  distancia por cada minuto se estima en 0.40 millas.

In [36]:
def filtrar_viajes_cortos_y_calcular_relacion(df):
    # Filtrar los viajes que duraron más de una hora
    df_viajes_cortos = df[df['Duracion'] <= 10]

    # Calcular la relación de distancia y tiempo
    df_viajes_cortos.loc[:,'Distancia_por_Minuto'] = df_viajes_cortos['trip_distance'] / df_viajes_cortos['Duracion']
    
    return df_viajes_cortos

df_viajes_cortos = filtrar_viajes_cortos_y_calcular_relacion(df[['trip_distance', 'Duracion', 'Distancia_por_Minuto']])
df_viajes_cortos.head()

trip_distance  Duracion  Distancia_por_Minuto
0            0.80  4.733333              0.169014
2            1.62  4.050000              0.400000
3            0.74  1.866667              0.396429
9            1.02  4.983333              0.204682
11           1.37  7.433333              0.184305

_________________________________________________________________________________________________________


En esta sección, he calculado la métrica correspondiente para determinar la distancia total recorrida por un taxi durante el transcurso de un mes.<br>

Creación de la Columna ‘Parte_del_Día’
Este código crea una columna ‘parte_del_dia’, que contiene la parte del día correspondiente a cada hora. La operación (horas+4)%24 es para que la noche comience a las 20:00 y termine a las 4:00 de la mañana.<br>

La función pd.cut() se utiliza para dividir las horas en los intervalos definidos por bins.

In [11]:
# Lectura del segundo dataframe 'fhvhv_tripdata_2023-09.parquet'.
df2=pd.read_parquet('../parquet/fhvhv_tripdata_2023-09.parquet')
df2.head(15) # Visualizacion de las primeras 15 filas.

hvfhs_license_num dispatching_base_num originating_base_num  \
0             HV0003               B03404               B03404   
1             HV0003               B03404               B03404   
2             HV0003               B03404               B03404   
3             HV0003               B03404               B03404   
4             HV0003               B03404               B03404   
5             HV0003               B03404               B03404   
6             HV0003               B03404               B03404   
7             HV0003               B03404               B03404   
8             HV0005               B03406                 None   
9             HV0005               B03406                 None   
10            HV0005               B03406                 None   
11            HV0003               B03404               B03404   
12            HV0003               B03404               B03404   
13            HV0003               B03404               B03404   
14            HV0003               B03404               B03404   

      request_datetime   on_scene_datetime     pickup_datetime  \
0  2023-09-01 00:00:13 2023-09-01 00:02:47 2023-09-01 00:02:56   
1  2023-09-01 00:27:24 2023-09-01 00:29:32 2023-09-01 00:29:44   
2  2023-09-01 00:25:14 2023-09-01 00:27:15 2023-09-01 00:28:01   
3  2023-09-01 00:31:44 2023-09-01 00:36:51 2023-09-01 00:36:51   
4  2023-09-01 00:08:33 2023-09-01 00:10:53 2023-09-01 00:12:28   
5  2023-09-01 00:26:26 2023-09-01 00:29:01 2023-09-01 00:31:01   
6  2023-09-01 00:06:09 2023-09-01 00:07:40 2023-09-01 00:08:32   
7  2023-09-01 00:56:51 2023-09-01 00:59:00 2023-09-01 00:59:57   
8  2023-09-01 00:07:31                 NaT 2023-09-01 00:11:46   
9  2023-09-01 00:28:32                 NaT 2023-09-01 00:36:42   
10 2023-09-01 00:40:04                 NaT 2023-09-01 00:44:37   
11 2023-09-01 00:22:53 2023-09-01 00:25:11 2023-09-01 00:27:12   
12 2023-09-01 00:42:55 2023-09-01 00:44:08 2023-09-01 00:46:07   
13 2023-09-01 00:23:43 2023-09-01 00:30:15 2023-09-01 00:31:04   
14 2023-08-31 23:47:30 2023-08-31 23:59:42 2023-09-01 00:00:13   

      dropoff_datetime  PULocationID  DOLocationID  trip_miles  ...  \
0  2023-09-01 00:14:08             7           226       2.350  ...   
1  2023-09-01 00:59:58           144            36       6.790  ...   
2  2023-09-01 00:35:53           186           125       1.610  ...   
3  2023-09-01 00:46:48           125           148       1.370  ...   
4  2023-09-01 00:27:47           170            48       1.290  ...   
5  2023-09-01 00:46:50           100           239       2.470  ...   
6  2023-09-01 00:24:17           141           238       3.540  ...   
7  2023-09-01 01:09:07            89           188       0.910  ...   
8  2023-09-01 00:29:56           255            61       3.808  ...   
9  2023-09-01 00:42:00           189           189       0.955  ...   
10 2023-09-01 00:54:36           189           188       1.491  ...   
11 2023-09-01 00:41:41           148           170       2.180  ...   
12 2023-09-01 01:18:27           170            49       7.310  ...   
13 2023-09-01 01:01:26           205            72      13.730  ...   
14 2023-09-01 00:34:18            93           265      20.380  ...   

    sales_tax  congestion_surcharge  airport_fee  tips  driver_pay  \
0        1.03                  0.00          0.0  0.00        9.41   
1        2.64                  2.75          0.0  7.18       25.97   
2        0.97                  2.75          0.0  0.00        8.34   
3        0.85                  0.75          0.0  0.00        7.44   
4        1.45                  2.75          0.0  0.00       13.30   
5        2.71                  2.75          0.0  5.29       17.70   
6        1.37                  2.75          0.0  0.00       15.83   
7        0.75                  0.00          0.0  0.00        6.45   
8        1.81                  0.00          0.0  0.00       15.38   
9        0.62                  0.00          0.0  0.

In [13]:
# Hago esta funcion para eliminar las columnas que no necesito.
def eliminar_columnas(df2):
    # Lista de columnas a eliminar
    columnas = ['shared_request_flag', 'shared_match_flag', 'access_a_ride_flag', 'wav_request_flag', 'wav_match_flag']
    
    # Eliminar las columnas
    df2 = df2.drop(columnas, axis=1)
    return df2

eliminar_columnas(df2)

hvfhs_license_num dispatching_base_num originating_base_num  \
0                   HV0003               B03404               B03404   
1                   HV0003               B03404               B03404   
2                   HV0003               B03404               B03404   
3                   HV0003               B03404               B03404   
4                   HV0003               B03404               B03404   
...                    ...                  ...                  ...   
19851118            HV0005               B03406                 None   
19851119            HV0003               B03404               B03404   
19851120            HV0003               B03404               B03404   
19851121            HV0005               B03406                 None   
19851122            HV0005               B03406                 None   

            request_datetime   on_scene_datetime     pickup_datetime  \
0        2023-09-01 00:00:13 2023-09-01 00:02:47 2023-09-01 00:02:56   
1        2023-09-01 00:27:24 2023-09-01 00:29:32 2023-09-01 00:29:44   
2        2023-09-01 00:25:14 2023-09-01 00:27:15 2023-09-01 00:28:01   
3        2023-09-01 00:31:44 2023-09-01 00:36:51 2023-09-01 00:36:51   
4        2023-09-01 00:08:33 2023-09-01 00:10:53 2023-09-01 00:12:28   
...                      ...                 ...                 ...   
19851118 2023-09-30 23:18:57                 NaT 2023-09-30 23:26:48   
19851119 2023-09-30 23:16:02 2023-09-30 23:22:48 2023-09-30 23:24:00   
19851120 2023-09-30 23:20:51 2023-09-30 23:41:52 2023-09-30 23:42:26   
19851121 2023-09-30 23:08:50                 NaT 2023-09-30 23:15:03   
19851122 2023-09-30 23:38:07                 NaT 2023-09-30 23:49:49   

            dropoff_datetime  PULocationID  DOLocationID  trip_miles  \
0        2023-09-01 00:14:08             7           226       2.350   
1        2023-09-01 00:59:58           144            36       6.790   
2        2023-09-01 00:35:53           186           125       1.610   
3        2023-09-01 00:46:48           125           148       1.370   
4        2023-09-01 00:27:47           170            48       1.290   
...                      ...           ...           ...         ...   
19851118 2023-09-30 23:59:21            79           246       3.461   
19851119 2023-09-30 23:38:33           181           148       3.120   
19851120 2023-10-01 00:03:49           148           265       4.270   
19851121 2023-09-30 23:30:43           200            20       4.003   
19851122 2023-10-01 00:03:51            47           159       2.978   

          trip_time  base_passenger_fare  tolls   bcf  sales_tax  \
0               672                11.57    0.0  0.32       1.03   
1              1814                29.72    0.0  0.82       2.64   
2               472                10.89    0.0  0.30       0.97   
3               597                 9.60    0.0  0.26       0.85   
4               919                16.29    0.0  0.45       1.45   
...             ...                  ...    ...   ...        ...   
19851118       1953                20.62    0.0  0.57       1.83   
19851119        873                23.44    0.0  0.64       2.08   
19851120       1283                31.56   20.0  1.42       0.00   
19851121        940                21.28    0.0  0.59       1.89   
19851122        842                16.95    0.0  0.47       1.51   

          congestion_surcharge  airport_fee  tips  driver_pay  
0                         0.00          0.0  0.00        9.41  
1                         2.75          0.0  7.18       25.97  
2                         2.75          0.0  0.00        8.34  
3                         0.75          0.0  0.00        7.44  
4                         2.75          0.0  0.00       13.30  
...                        ...          ...   ...         ...  
19851118                  2.75          0.0  0.00       26.12  
19851119                  2.75          0.0  0.00       14.56  
19851120             

Este codigo crea una columna 'parte_del_dia', que contiene la parte del dia correspondiente a cada hora. La operacion (horas+4)%24 es para que la noche comience a las 20:00 y termine a las 4:00 de la mañana.<br>

La impresion del codigo muestra los viajes realizados por mes.

La funcion pd.cut()se utiliza para dividir las horas en los intervalos definidos por bins

In [14]:
df2['pickup_datetime'] = pd.to_datetime(df2['pickup_datetime'])

# Creo una columna para la parte del día
horas = df2['pickup_datetime'].dt.hour # Extraer las horas
bins = [0, 4, 12, 20, 24] #| 0-4: Noche | 4-12: Mañana | 12-20: Tarde | 20-24: Noche
etiquetas = ['Noche', 'Mañana', 'Tarde', 'Noche']
# Crear la columna 'parte_del_dia'
# Agregar 4 horas para que coincida con el huso horario de Nueva York
df2['parte_del_dia'] = pd.cut((horas+4)%24, bins=bins, labels=etiquetas, ordered=False) 
# Calculando la demanda de viajes para cada parte del día
demanda = df2['parte_del_dia'].value_counts()

# Formatear los números con comas como separadores de miles
demanda_format = demanda.apply(lambda x: "{:,}".format(x)) 

demanda_format


parte_del_dia
Tarde     7,436,604
Noche     7,207,792
Mañana    4,109,288
Name: count, dtype: object

In [15]:
df2.shape # Muestra el numero de filas y columnas del dataframe.

(19851123, 25)

_________________________________________________________________________________________________
El bloque de código realiza una serie de operaciones destinadas a analizar y presentar de manera clara las ganancias generadas por los taxis a lo largo del día. Se crea una nueva columna que resume diversos componentes del costo del viaje en taxi. Luego, se agrupan y suman estas ganancias por partes del día, y los resultados se formatean para facilitar la lectura. La visualización final proporciona una perspectiva detallada de cómo las ganancias de los taxis varían en diferentes momentos del día.

In [16]:
# Crear una nueva columna 'total_ganancias_taxi' que es la suma de las columnas de costo
df2['total_ganancias_taxi'] = df2[['base_passenger_fare', 'tolls', 'bcf', 'sales_tax', 'congestion_surcharge', 'airport_fee', 'tips']].sum(axis=1)

# Calcular el dinero generado en cada parte del día
dinero_generado = df2.groupby('parte_del_dia', observed=True)['total_ganancias_taxi'].sum()
# Formatear los números con comas como separadores de miles
dinero_generado_format = dinero_generado.apply(lambda x: "{:,.2f}".format(x))

dinero_generado_format

parte_del_dia
Mañana    132,577,174.34
Noche     232,532,129.34
Tarde     248,147,368.95
Name: total_ganancias_taxi, dtype: object

_______________________________________________________________________________________________________
La función desarrollada, se ha logrado calcular la distancia en millas y las emisiones de dióxido de carbono por viaje. La función extrae y utiliza las columnas relevantes del DataFrame, agregando información  para comprender la relación entre la distancia recorrida en millas y las emisiones de dióxido de carbono asociadas a cada viaje. 

In [8]:
df3 = pd.read_csv('../csv/MY2023 Fuel Consumption Ratings.csv', encoding='ISO-8859-1', low_memory=False)
df3.head(5) # Visualizacion de las primeras 5 filas.

Model   Make         Model.1 Vehicle Class Engine Size  Cylinders  \
0  Year    NaN             NaN           NaN         (L)        NaN   
1  2023  Acura         Integra     Full-size         1.5        4.0   
2  2023  Acura  Integra A-SPEC     Full-size         1.5        4.0   
3  2023  Acura  Integra A-SPEC     Full-size         1.5        4.0   
4  2023  Acura      MDX SH-AWD    SUV: Small         3.5        6.0   

  Transmission  Fuel Fuel Consumption      Unnamed: 9  ... Unnamed: 214  \
0          NaN  Type  City (L/100 km)  Hwy (L/100 km)  ...          NaN   
1          AV7     Z              7.9             6.3  ...          NaN   
2          AV7     Z              8.1             6.5  ...          NaN   
3           M6     Z              8.9             6.5  ...          NaN   
4         AS10     Z             12.6             9.4  ...          NaN   

  Unnamed: 215 Unnamed: 216 Unnamed: 217 Unnamed: 218  Unnamed: 219  \
0          NaN          NaN          NaN          NaN           NaN   
1          NaN          NaN          NaN          NaN           NaN   
2          NaN          NaN          NaN          NaN           NaN   
3          NaN          NaN          NaN          NaN           NaN   
4          NaN          NaN          NaN          NaN           NaN   

   Unnamed: 220  Unnamed: 221  Unnamed: 222  Unnamed: 223  
0           NaN           NaN           NaN           NaN  
1           NaN           NaN           NaN           NaN  
2           NaN           NaN           NaN           NaN  
3           NaN           NaN           NaN           NaN  
4           NaN           NaN           NaN           NaN  

[5 rows x 224 columns]

In [9]:
# Visualizacion de las c olumnas 'CO2 Emissions','CO2','Smog'.
df3[['CO2 Emissions','CO2','Smog']].head()

CO2 Emissions     CO2    Smog
0        (g/km)  Rating  Rating
1           167       6       7
2           172       6       7
3           181       6       6
4           263       4       5

In [118]:
df_unido = pd.DataFrame({
    'trip_distance': df['trip_distance'],
    'Duracion': df['Duracion'],
    'Distancia_por_Minuto': df['Distancia_por_Minuto'],
    'passenger_count': df['passenger_count'],
    'total_amount': df['total_amount'],
    'parte_del_dia': df2['parte_del_dia'],
    'CO2 Emissions': df3['CO2 Emissions'],
    'CO2': df3['CO2'],
    'Smog': df3['Smog'],
}) # Creo un nuevo dataframe con las columnas que necesito.

In [114]:
df_unido.head() #| Visualizacion de las primeras 5 filas del nuevo dataframe.

trip_distance        Duracion  Distancia_por_Minuto  passenger_count  \
0           0.80 0 days 00:04:44              0.169014              1.0   
1           2.34 0 days 00:11:48              0.198305              2.0   
2           1.62 0 days 00:04:03              0.400000              1.0   
3           0.74 0 days 00:01:52              0.396429              1.0   
4           9.85 0 days 00:36:42              0.268392              1.0   

   total_amount parte_del_dia CO2 Emissions     CO2    Smog  
0         11.50         Noche        (g/km)  Rating  Rating  
1         21.20         Noche           167       6       7  
2         15.60         Noche           172       6       7  
3         11.10         Noche           181       6       6  
4         73.77         Noche           263       4       5

In [115]:
# visualizacion de los tipos de datos de las columnas.
print(df_unido['trip_distance'].dtype) 
print(df_unido['CO2 Emissions'].dtype)

float64
object


In [20]:
# Cambio el tipo de dato de las columnas 'trip_distance' y 'CO2 Emissions' a numerico.
df_unido['trip_distance'] = pd.to_numeric(df_unido['trip_distance'], errors='coerce')
df_unido['CO2 Emissions'] = pd.to_numeric(df_unido['CO2 Emissions'], errors='coerce')

In [111]:
# Funcion para calcular las emisiones de CO2 por viaje.
def calcular_emisiones(df_unido):
    # Calcular las emisiones de CO2 por viaje
    df_unido['CO2_emissions_por_viaje'] = df_unido['trip_distance'] * df_unido['CO2 Emissions'] / 1000 #| 1 km = 0.621371 millas

    return df_unido

df_unido = calcular_emisiones(df_unido)
df_unido[['trip_distance', 'CO2_emissions_por_viaje']].head()

trip_distance  CO2_emissions_por_viaje
0           0.80                      NaN
1           2.34                  0.39078
2           1.62                  0.27864
3           0.74                  0.13394
4           9.85                  2.59055

Creo un nuevo CSV llamado 'Taxis_Amarillos_CO2_NY' esto me ayudara con la siguiente fase la EDA.

In [116]:
df_unido.to_csv('Taxis_Amarillos_CO2_NY.csv',index=False) 

- Hago un ETL y lectura del archivo "green_tripdata_2023-09.parquet"<br>

Los taxis verdes como a los taxis amarillos, tienen reglas similares en términos de compartir viajes. Pueden recoger a múltiples pasajeros en un solo viaje si tienen destinos cercanos o en la misma dirección.


In [79]:
df4=pd.read_parquet('../parquet/green_tripdata_2023-09.parquet')
df4.head(2000) # Visualizacion de las primeras 15 filas.

VendorID lpep_pickup_datetime lpep_dropoff_datetime store_and_fwd_flag  \
0            1  2023-09-01 00:08:04   2023-09-01 00:13:39                  N   
1            1  2023-09-01 00:30:44   2023-09-01 00:39:30                  N   
2            2  2023-09-01 00:14:39   2023-09-01 00:38:30                  N   
3            2  2023-09-01 00:35:48   2023-09-01 00:56:48                  N   
4            2  2023-09-01 00:32:03   2023-09-01 00:43:11                  N   
...        ...                  ...                   ...                ...   
1995         2  2023-09-01 22:42:24   2023-09-01 22:52:26                  N   
1996         2  2023-09-01 22:38:05   2023-09-01 22:52:51                  N   
1997         2  2023-09-01 22:54:49   2023-09-01 23:13:19                  N   
1998         2  2023-09-01 22:23:50   2023-09-01 22:23:55                  N   
1999         2  2023-09-01 22:11:45   2023-09-01 22:36:38                  N   

      RatecodeID  PULocationID  DOLocationID  passenger_count  trip_distance  \
0            1.0            41            75              1.0           1.10   
1            1.0            74           236              1.0           1.80   
2            5.0            93           205              2.0          10.52   
3            1.0           112            17              1.0           4.72   
4            1.0           131            82              1.0           5.95   
...          ...           ...           ...              ...            ...   
1995         1.0           129           223              1.0           2.52   
1996         1.0            61           188              1.0           2.33   
1997         1.0            33            17              1.0           2.94   
1998         5.0           219           219              2.0           0.00   
1999         1.0            93           198              2.0           6.66   

      fare_amount  extra  mta_tax  tip_amount  tolls_amount  ehail_fee  \
0             7.9   1.00      1.5        0.00           0.0        NaN   
1            10.7   3.75      1.5        3.15           0.0        NaN   
2            60.0   0.00      0.0        0.00           0.0        NaN   
3            23.3   1.00      0.5       10.00           0.0        NaN   
4            24.0   1.00      0.5        0.00           0.0        NaN   
...           ...    ...      ...         ...           ...        ...   
1995         14.2   1.00      0.5        4.18           0.0        NaN   
1996         14.9   1.00      0.5        0.00           0.0        NaN   
1997         19.1   1.00      0.5        0.00           0.0        NaN   
1998         85.0   0.00      0.0       15.00           0.0        NaN   
1999         32.4   1.00      0.5        0.00           0.0        NaN   

      improvement_surcharge  total_amount  payment_type  trip_type  \
0                       1.0         10.40           1.0        1.0   
1                       1.0         19.10           1.0        1.0   
2                       1.0         61.00           1.0        2.0   
3                       1.0         35.80           1.0        1.0   
4                       1.0         26.50           2.0        1.0   
...                     ...           ...           ...        ...   
1995                    1.0         20.88           1.0        1.0   
1996                    1.0         17.40           2.0        1.0   
1997                    1.0         21.60           2.0        1.0   
1998                    1.0        101.00           1.0        2.0   
1999                    1.0         34.90           2.0        1.0   

      congestion_surcharge  
0                     0.00  
1                     2.75  
2                     0.00  
3                     0.00  
4                     0.00  
...                    ...  
1995                  0.00  
1996                  0.00  
1997                  0.00  
1998                  0.00  
1999                  0.00  

[2000

In [59]:
df4.shape # Muestra el numero de filas y columnas del dataframe.

(65471, 22)

In [80]:
# Hago esta funcion para contar los datos faltantes en cada columna.
def contar_datos_faltantes(df4):
    # Contar el número de datos faltantes en cada columna
    datos_faltantes = df4.isnull().sum()
    return datos_faltantes

contar_datos_faltantes(df4)

VendorID                     0
lpep_pickup_datetime         0
lpep_dropoff_datetime        0
store_and_fwd_flag        4615
RatecodeID                4615
PULocationID                 0
DOLocationID                 0
passenger_count           4615
trip_distance                0
fare_amount                  0
extra                        0
mta_tax                      0
tip_amount                   0
tolls_amount                 0
ehail_fee                65471
improvement_surcharge        0
total_amount                 0
payment_type              4615
trip_type                 4621
congestion_surcharge      4615
dtype: int64

In [82]:
# Hago esta funcion para calcular la duracion del viaje.
def calcular_duracion(df4, lpep_pickup_datetime, lpep_dropoff_datetime ):
    
    # Uso la funcion to_datetime para convertir las columnas en formato fecha.
    df4[lpep_pickup_datetime] = pd.to_datetime(df4[lpep_pickup_datetime])
    df4[lpep_dropoff_datetime ] = pd.to_datetime(df4[lpep_dropoff_datetime ])
    
    # Calcular la duración
    df4['Duracion'] = df4[lpep_dropoff_datetime ] - df4[lpep_pickup_datetime]
    
    return df4

calcular_duracion(df4, 'lpep_pickup_datetime', 'lpep_dropoff_datetime')

VendorID lpep_pickup_datetime lpep_dropoff_datetime store_and_fwd_flag  \
0             1  2023-09-01 00:08:04   2023-09-01 00:13:39                  N   
1             1  2023-09-01 00:30:44   2023-09-01 00:39:30                  N   
2             2  2023-09-01 00:14:39   2023-09-01 00:38:30                  N   
3             2  2023-09-01 00:35:48   2023-09-01 00:56:48                  N   
4             2  2023-09-01 00:32:03   2023-09-01 00:43:11                  N   
...         ...                  ...                   ...                ...   
65466         2  2023-09-30 22:09:00   2023-09-30 22:21:00               None   
65467         2  2023-09-30 22:06:00   2023-09-30 22:15:00               None   
65468         2  2023-09-30 22:02:00   2023-09-30 22:38:00               None   
65469         2  2023-09-30 23:22:00   2023-09-30 23:38:00               None   
65470         2  2023-09-30 23:17:00   2023-09-30 23:42:00               None   

       RatecodeID  PULocationID  DOLocationID  passenger_count  trip_distance  \
0             1.0            41            75              1.0           1.10   
1             1.0            74           236              1.0           1.80   
2             5.0            93           205              2.0          10.52   
3             1.0           112            17              1.0           4.72   
4             1.0           131            82              1.0           5.95   
...           ...           ...           ...              ...            ...   
65466         NaN            17            35              NaN           2.19   
65467         NaN           166           244              NaN           2.51   
65468         NaN           166           132              NaN        6452.20   
65469         NaN           159           137              NaN           7.39   
65470         NaN            66            61              NaN           3.64   

       fare_amount  ...  mta_tax  tip_amount  tolls_amount  ehail_fee  \
0             7.90  ...      1.5        0.00          0.00        NaN   
1            10.70  ...      1.5        3.15          0.00        NaN   
2            60.00  ...      0.0        0.00          0.00        NaN   
3            23.30  ...      0.5       10.00          0.00        NaN   
4            24.00  ...      0.5        0.00          0.00        NaN   
...            ...  ...      ...         ...           ...        ...   
65466        12.53  ...      0.0        2.44          0.00        NaN   
65467        15.88  ...      0.0        3.38          0.00        NaN   
65468        57.05  ...      0.0        9.75          6.94        NaN   
65469        29.61  ...      0.0        0.00          0.00        NaN   
65470        24.25  ...      0.0        5.05          0.00        NaN   

       improvement_surcharge  total_amount  payment_type  trip_type  \
0                        1.0         10.40           1.0        1.0   
1                        1.0         19.10           1.0        1.0   
2                        1.0         61.00           1.0        2.0   
3                        1.0         35.80           1.0        1.0   
4                        1.0         26.50           2.0        1.0   
...                      ...           ...           ...        ...   
65466                    1.0         15.97           NaN        NaN   
65467                    1.0         20.26           NaN        NaN   
65468                    1.0         74.74           NaN        NaN   
65469                    1.0         33.36           NaN        NaN   
65470                    1.0         30.30           NaN        NaN   

       congestion_surcharge        Duracion  
0                      0.00 0 days 00:05:35  
1                      2.75 0 days 00:08:46  
2                      0.00 0 days 00:23:51  
3                      0.00 0 days 00:21:00  
4                      0.00 0 days 00:11:08  
...                     ...             ...  
65466    

In [84]:
# Seleccionar solo las columnas 'lpep_pickup_datetime', 'lpep_dropoff_datetime' y 'Duracion'
df_seleccionado = df4[['lpep_pickup_datetime', 'lpep_dropoff_datetime','trip_distance','Duracion']]

df_seleccionado.head()

lpep_pickup_datetime lpep_dropoff_datetime  trip_distance        Duracion
0  2023-09-01 00:08:04   2023-09-01 00:13:39           1.10 0 days 00:05:35
1  2023-09-01 00:30:44   2023-09-01 00:39:30           1.80 0 days 00:08:46
2  2023-09-01 00:14:39   2023-09-01 00:38:30          10.52 0 days 00:23:51
3  2023-09-01 00:35:48   2023-09-01 00:56:48           4.72 0 days 00:21:00
4  2023-09-01 00:32:03   2023-09-01 00:43:11           5.95 0 days 00:11:08

_________________________________________________________________________________

In [96]:
df4['lpep_pickup_datetime'] = pd.to_datetime(df4['lpep_pickup_datetime'])

# Creo una columna para la parte del día
horas = df4['lpep_pickup_datetime'].dt.hour # Extraer las horas
bins = [0, 4, 12, 20, 24] #| 0-4: Noche | 4-12: Mañana | 12-20: Tarde | 20-24: Noche
etiquetas = ['Noche', 'Mañana', 'Tarde', 'Noche']
# Crear la columna 'viajes_todo_mes'
# Agregar 4 horas para que coincida con el huso horario de Nueva York
df4['viajes_todo_mes'] = pd.cut((horas+4)%24, bins=bins, labels=etiquetas, ordered=False) 
# Calculando la demanda de viajes para cada parte del día
demanda = df4['viajes_todo_mes'].value_counts()

# Formatear los números con comas como separadores de miles
demanda_format = demanda.apply(lambda x: "{:,}".format(x)) 

demanda_format


viajes_todo_mes
Tarde     30,472
Noche     22,520
Mañana     9,299
Name: count, dtype: object

 La función realiza una operación de agregación para calcular las ganancias mensuales a partir de las  transacciones de los taxis verdes.<br>

 La presencia de datos en el año 2009 en la salida parece ser un error o un dato atípico, ya que generalmente se espera que los datos se centren en el rango de fechas más recientes de septiembre 2023.

In [98]:
def calcular_ganancias_mensuales(df4):
    # Asegurarse de que 'lpep_pickup_datetime' es de tipo datetime
    if df4['lpep_pickup_datetime'].dtype == 'object':
        df4['lpep_pickup_datetime'] = pd.to_datetime(df4['lpep_pickup_datetime'])
    
     # Filtrar solo los datos del año 2023
    df4_2023 = df4[df4['lpep_pickup_datetime'].dt.year == 2023]
    # Extraer el mes y el año de 'lpep_pickup_datetime'
    df4['Mes'] = df4['lpep_pickup_datetime'].dt.to_period('M')
    
    # Calcular las ganancias mensuales
    ganancias_mensuales = df4.groupby('Mes')['total_amount'].sum()

    return ganancias_mensuales

ganancias_mensuales = calcular_ganancias_mensuales(df4)
calcular_ganancias_mensuales(df4)

Mes
2009-01         29.55
2023-08       1896.38
2023-09    1736710.91
2023-10         31.13
Freq: M, Name: total_amount, dtype: float64

In [99]:
df4.head()

VendorID lpep_pickup_datetime lpep_dropoff_datetime store_and_fwd_flag  \
0         1  2023-09-01 00:08:04   2023-09-01 00:13:39                  N   
1         1  2023-09-01 00:30:44   2023-09-01 00:39:30                  N   
2         2  2023-09-01 00:14:39   2023-09-01 00:38:30                  N   
3         2  2023-09-01 00:35:48   2023-09-01 00:56:48                  N   
4         2  2023-09-01 00:32:03   2023-09-01 00:43:11                  N   

   RatecodeID  PULocationID  DOLocationID  passenger_count  trip_distance  \
0         1.0            41            75              1.0           1.10   
1         1.0            74           236              1.0           1.80   
2         5.0            93           205              2.0          10.52   
3         1.0           112            17              1.0           4.72   
4         1.0           131            82              1.0           5.95   

   fare_amount  ...  total_amount  payment_type  trip_type  \
0          7.9  ...          10.4           1.0        1.0   
1         10.7  ...          19.1           1.0        1.0   
2         60.0  ...          61.0           1.0        2.0   
3         23.3  ...          35.8           1.0        1.0   
4         24.0  ...          26.5           2.0        1.0   

   congestion_surcharge        Duracion  parte_del_dia         dia      mes  \
0                  0.00 0 days 00:05:35          Noche  2023-09-01  2023-09   
1                  2.75 0 days 00:08:46          Noche  2023-09-01  2023-09   
2                  0.00 0 days 00:23:51          Noche  2023-09-01  2023-09   
3                  0.00 0 days 00:21:00          Noche  2023-09-01  2023-09   
4                  0.00 0 days 00:11:08          Noche  2023-09-01  2023-09   

   viajes_todo_mes      Mes  
0            Noche  2023-09  
1            Noche  2023-09  
2            Noche  2023-09  
3            Noche  2023-09  
4            Noche  2023-09  

[5 rows x 26 columns]

In [102]:
df_unido2 = pd.DataFrame({
    'trip_distance': df4['trip_distance'],
    'CO2 Emissions': df3['CO2 Emissions'],
    'CO2': df3['CO2'],
    'Smog': df3['Smog'],
    'passenger_count': df4['passenger_count'],
    'total_amount': df4['total_amount'],
    'parte_del_dia': df4['parte_del_dia'],
    'dia': df4['dia'],
    'mes': df4['mes'],
    'viajes_todo_mes': df4['viajes_todo_mes'],
    'Mes': df4['Mes'],
    'Duracion': df4['Duracion']
    
}) # Creo un nuevo dataframe con las columnas que necesito.

In [103]:
df_unido2.head() #| Visualizacion de las primeras 5 filas del nuevo dataframe.

trip_distance CO2 Emissions     CO2    Smog  passenger_count  total_amount  \
0           1.10        (g/km)  Rating  Rating              1.0          10.4   
1           1.80           167       6       7              1.0          19.1   
2          10.52           172       6       7              2.0          61.0   
3           4.72           181       6       6              1.0          35.8   
4           5.95           263       4       5              1.0          26.5   

  parte_del_dia         dia      mes viajes_todo_mes      Mes        Duracion  
0         Noche  2023-09-01  2023-09           Noche  2023-09 0 days 00:05:35  
1         Noche  2023-09-01  2023-09           Noche  2023-09 0 days 00:08:46  
2         Noche  2023-09-01  2023-09           Noche  2023-09 0 days 00:23:51  
3         Noche  2023-09-01  2023-09           Noche  2023-09 0 days 00:21:00  
4         Noche  2023-09-01  2023-09           Noche  2023-09 0 days 00:11:08

In [67]:
print(df_unido2[['trip_distance', 'CO2 Emissions']].dtypes)

trip_distance    float64
CO2 Emissions     object
dtype: object


In [104]:
# Cambio el tipo de dato de las columnas 'trip_distance' y 'CO2 Emissions' a numerico.
df_unido2['trip_distance'] = pd.to_numeric(df_unido2['trip_distance'], errors='coerce')
df_unido2['CO2 Emissions'] = pd.to_numeric(df_unido2['CO2 Emissions'], errors='coerce')

In [110]:
# Funcion para calcular las emisiones de CO2 por viaje.
def calcular_emisiones(df_unido2):
    # Calcular las emisiones de CO2 por viaje
    df_unido2['CO2_emissions_por_viaje'] = df_unido2['trip_distance'] * df_unido2['CO2 Emissions'] / 1000 #| 1 km = 0.621371 millas

    return df_unido2

df_unido2 = calcular_emisiones(df_unido2)
df_unido2[['trip_distance', 'CO2_emissions_por_viaje']].head()

trip_distance  CO2_emissions_por_viaje
0           1.10                      NaN
1           1.80                  0.30060
2          10.52                  1.80944
3           4.72                  0.85432
4           5.95                  1.56485

In [112]:
df_unido2.to_csv('Taxis_Verdes_CO2_NY.csv',index=False) #| Guardo el dataframe en un archivo csv.